### ISSUE TITLE
Cache Google Container Registry Repositories

### ISSUE THREAD

**Issue by JXavierMSFT** [Azure/acr#686]
Caching for ACR will soon allow users to cache repositories from Google Container Registry. This functionality is tentatively scheduled for release in late July.

**Comment by mac2000** [comment]
Wondering if understand that one correct: in Google we have quadrillion images, aka `gcr.io/myproject/myapp1`, we wish to migrate to Azure, to make things smoother, idea is to configure Azure as a cache first, aka `myproject.azurecr.io/myapp1` will be retrieved from Google via Azure and cached, so we can start changing images everywhere to fill up the case, and later, instead of cached versions we will just push new versions directly to Azure instead of Google

**Comment by mabenedi** [comment]
Yes, but with two caveats. 

1. Caching works at the repo level. You will need an active cache rule for each repo you want to cache. So if you have 50 repos in GCR that you want to cache in ACR, you'll need to create 50 cache rules (this is something you'll probably want to write a script for when the time comes). Note that we currently have a limit of 50 cache rules, although I think we're planning on raising this in the future. 

2. When you have an active cache rule associated with a repo in ACR, you cannot manually push new images and tags to that repo. So you would have to delete the cache rule prior to pushing new images the repo. All images that had been cached prior to the deletion of the cache rule would remain in the repo.

**Comment by mac2000** [comment]
Still better than nothing but I bed 50 rules limit is the one we will face immediately 🤷‍♂️

but because we are going to script things in either case nothing prevents us just to pull push things manually

which seems to be workaround so far

**Comment by JXavierMSFT** [comment]
Hello Everyone, 

Unfortunately, we won't be able to ship GCR as an available upstream this month. I will update this thread when I have a new accurate Timeline.

**Comment by frank-m** [comment]
Hi, is there any update on the new timeline?

**Comment by mabenedi** [comment]
Yes, the PRs necessary to enable gcr.io as an upstream have been merged and will be shipped on our next deployment, within a month. We can post again here when the deployment is completed in all regions

**Comment by sengjea** [comment]
I tried running `az acr Cache create -r tmazureemearegistrycr -n sengjea-test -s 'eu.gcr.io/my-root-project/*' -t 'sync'` and I seem to get ```(UpstreamNotSupported) Unsupported upstream or login server 'eu.gcr.io/tm-root-project/*' provided. Supported upstreams or login servers are: docker.io, mcr.microsoft.com, quay.io, public.ecr.aws, ghcr.io, nvcr.io, registry.k8s.io, gcr.io.```

Is the old hostname not supported?

**Comment by sengjea** [comment]
I ran `az acr Cache create -r myazurecr -n myrulename -s 'eu.gcr.io/my-gcr-registry/*' -t 'sync/*'`
```
(UpstreamNotSupported) Unsupported upstream or login server 'eu.gcr.io/my-gcr-registry/*' provided. Supported upstreams or login servers are: docker.io, mcr.microsoft.com, quay.io, public.ecr.aws, ghcr.io, nvcr.io, registry.k8s.io, gcr.io.
```

Running it with `gcr.io` even though the wrong hostname works. Is `eu.gcr.io` not going to be supported?

**Comment by kimsandb** [comment]
The docs at aka.ms/acr/cache claim that google cloud is supported with both authenticated pulls and unauthenticated pulls.

google cloud is supporting _json_key for authentication. I did not find any documentation how to setup authenticated pulls here.

I tried adding _json_key as username and privatekey as password but didn't work directly.
Is this supposed to work how?

**Comment by mabenedi** [comment]
> Running it with `gcr.io`, even though the wrong hostname, works. Is `eu.gcr.io` not going to be supported?

@sengjea We are looking at adding eu.gcr.io support.

**Comment by mabenedi** [comment]
> The docs at aka.ms/acr/cache claim that google cloud is supported with both authenticated pulls and unauthenticated pulls.
> 
> google cloud is supporting _json_key for authentication. I did not find any documentation how to setup authenticated pulls here.
> 
> I tried adding _json_key as username and privatekey as password but didn't work directly. Is this supposed to work how?

Yes, this is supported. `_json_key` should be your username secret value and the contents of the keyfile should be your password secret value. When we go to do authenticated pulls, we are just grabbing the values of these secrets and using them to do what is essentially a `docker login`. Theoretically, if it works for docker login, it _should_ work for authenticated pulls

**Comment by KevinDW-Fluxys** [comment]
I have some images which we want to pull from **-docker.pkg.dev which is google cloud registry but they are not allowed (via the CLI) as the domain is not gcr.io
Is there a way to make this work as well?

**Comment by mabenedi** [comment]
> @mabenedi I have some images which we want to pull from **-docker.pkg.dev which is google cloud registry but they are not allowed (via the CLI) as the domain is not gcr.io Is there a way to make this work as well?

There's not currently a way to make a cacherule with that as a source as we are strict about the domain allowlist regardless of the backing registry. 

I've passed your request along to the team to see if we can add `**-docker.pkg.dev` to the allowlist in the future

**Comment by Cognimity** [comment]
I am also trying to find relevant information on how to do the mirroring from gcr.io, but I can't find anywhere relevant documentation on how it's done.

The only thing provided on the documentation page is that the google container registry is supported, but it is not clear how this is done...

**Comment by IvanJosipovic** [comment]
Are there plans to allow the cache to work with other hostnames?

Unsupported upstream or login server 'xpkg.upbound.io/*' provided. Supported upstreams or login servers are: docker.io, mcr.microsoft.com, quay.io, public.ecr.aws, ghcr.io, nvcr.io, registry.k8s.io, gcr.io, eu.gcr.io.

**Comment by JXavierMSFT** [comment]
@IvanJosipovic Yes, we are looking into adding more upstreams. Do you have any in mind that you'd like to see?

**Comment by JXavierMSFT** [comment]
@Cognimity My apologies for the confusion you are experiencing with the Documentation considering Google Container Registry. I'll get these documents updated.

**Comment by IvanJosipovic** [comment]
> @IvanJosipovic Yes, we are looking into adding more upstreams. Do you have any in mind that you'd like to see?

The one I'm attempting currently is, "xpkg.upbound.io" for Crossplane, I will add more as I hit them.

Why is it even bound to specific hostnames, since the container image API is a standard protocol?

**Comment by IvanJosipovic** [comment]
@JXavierMSFT any update on adding xpkg.upbound.io?

**Comment by torbendury** [comment]
> [@mabenedi](https://github.com/mabenedi) 
> I've passed your request along to the team to see if we can add `**-docker.pkg.dev` to the allowlist in the future

Did the support for the new Google Artifact Registry feature bear any fruits? The team I'm working with is highly interested in such a feature, since the Google Container Registry will be decommissioned soon.

**Comment by mabenedi** [comment]
@toddysm @JXavierMSFT might be able to help, I'm no longer on the team

**Comment by torbendury** [comment]
Hello @toddysm @JXavierMSFT, do you have any news on this? I have migrated to Google Artifact Registry since it's read-only from 18th March on and will be shutdown in May, but now I'm unable to use ACR as a pull-through cache.

A valid option, IMHO, would be to support a more general "OCI Registry" upstream (as which a Google Artifact Registry and probably all other from the [list of supported registries](https://learn.microsoft.com/en-us/azure/container-registry/artifact-cache-overview#upstream-support) can act) that takes an endpoint as a minimal configuration and optionally user/password.

**Comment by JXavierMSFT** [comment]
Hey @torbendury, 

We are currently working on support for Google Artifact Registry. However, your request for supported a more general OCI registry support is currently being considered. However, that feature is currently in the planning stage of development.

**Comment by torbendury** [comment]
This sounds actually great! Is there any kind of public roadmap where we can get updates in the meantime, or the roughest ETA on the Artifact Registry support?

Best regards

**Comment by Justin-DynamicD** [comment]
> Hey [@torbendury](https://github.com/torbendury),
> 
> We are currently working on support for Google Artifact Registry. However, your request for supported a more general OCI registry support is currently being considered. However, that feature is currently in the planning stage of development.

This is good to know, but should be worth noting that GCR is officially retired now, meaning many clients are simply cut off at this time.

Does there happen to be an ETA?

**Comment by cesanchezmsft** [comment]
> > Hey [@torbendury](https://github.com/torbendury),
> > We are currently working on support for Google Artifact Registry. However, your request for supported a more general OCI registry support is currently being considered. However, that feature is currently in the planning stage of development.
> 
> This is good to know, but should be worth noting that GCR is officially retired now, meaning many clients are simply cut off at this time.
> 
> Does there happen to be an ETA?

Hi @Justin-DynamicD, I wanted to let you know that we are currently working on adding the domain 'pkg.dev' to the allowed domains list for creating credentials, as well as setting up a cache rule for the new Google Artifact Registry.

Current error: 
(UpstreamNotSupported) Unsupported upstream or login server 'us-docker.pkg.dev' provided. Supported upstreams or login servers are: docker.io, mcr.microsoft.com, quay.io, public.ecr.aws, ghcr.io, nvcr.io, registry.k8s.io, gcr.io, eu.gcr.io.

We currently do not have an ETA for task completion. We apologize for any inconvenience this may cause.

**Comment by toddysm** [comment]
@mohammadidinani apologies about the delay - this fell through during the holiday season in US. Do you still experience the issue?
